src/dbnode/storage/bootstrap/bootstrapper/fs/source_instrumentation.go

```diff
@@ -0,0 +1,166 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fs
+
+import (
+	"time"
+
+	"github.com/opentracing/opentracing-go"
+	opentracinglog "github.com/opentracing/opentracing-go/log"
+	"github.com/uber-go/tally"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+
+	"github.com/m3db/m3/src/dbnode/tracepoint"
+	"github.com/m3db/m3/src/x/clock"
+	"github.com/m3db/m3/src/x/context"
+	"github.com/m3db/m3/src/x/instrument"
+)
+
+type instrumentationContext struct {
+	opts                   Options
+	nowFn                  clock.NowFn
+	log                    *zap.Logger
+	start                  time.Time
+	span                   opentracing.Span
+	bootstrapDataDuration  tally.Timer
+	bootstrapIndexDuration tally.Timer
+	profiler               instrument.Profiler
+	logFields              []zapcore.Field
+}
+
+func newInstrumentationContext(
+	opts Options,
+	nowFn clock.NowFn,
+	log *zap.Logger,
+	span opentracing.Span,
+	scope tally.Scope,
+	profiler instrument.Profiler,
+) *instrumentationContext {
+	return &instrumentationContext{
+		opts:                   opts,
+		nowFn:                  nowFn,
+		log:                    log,
+		span:                   span,
+		profiler:               profiler,
+		bootstrapDataDuration:  scope.Timer("data-duration"),
+		bootstrapIndexDuration: scope.Timer("index-duration"),
+		logFields: []zapcore.Field{
+			zap.Stringer("cachePolicy", opts.ResultOptions().SeriesCachePolicy()),
+		},
+	}
+}
+
+const (
+	dataProfile  = "fs-data"
+	indexProfile = "fs-index"
+)
+
+func (i *instrumentationContext) finish() {
+	i.span.Finish()
+}
+
+func (i *instrumentationContext) startCPUProfile(name string) {
+	err := i.profiler.StartCPUProfile(name)
+	if err != nil {
+		i.log.Error("unable to start cpu profile", zap.Error(err))
+	}
+}
+
+func (i *instrumentationContext) stopCPUProfile() {
+	if err := i.profiler.StopCPUProfile(); err != nil {
+		i.log.Error("unable to stop cpu profile", zap.Error(err))
+	}
+}
+
+func (i *instrumentationContext) writeHeapProfile(name string) {
+	err := i.profiler.WriteHeapProfile(name)
+	if err != nil {
+		i.log.Error("unable to write heap profile", zap.Error(err))
+	}
+}
+
+func (i *instrumentationContext) bootstrapDataStarted() {
+	i.log.Info("bootstrapping time series data start", i.logFields...)
+	i.span.LogFields(opentracinglog.String("event", "bootstrap_data_start"))
+	i.start = i.nowFn()
+	i.startCPUProfile(dataProfile)
+	i.writeHeapProfile(dataProfile)
```

Comment:
Let's make this either environment variable or config configurable and merge this PR.

i.e. we don't always want to run the profiler, having the code merged would be ideal but we don't want to run it every single bootstrap (it should be opt in).

Reply:
It uses no-op profiler by default **(https://github.com/m3db/m3/pull/3076/files#diff-c3d9d36368a991782e676e4918cd4d05819184ebb52db6e975af061fffd8fd85R38)** so all these methods will do nothing by default unless custom profiler implementation is injected.